Added Huber loss

Sentient07 opened the pull request:
For #814

f0k commented on lasagne/objectives.py:
```diff
@@ -338,6 +338,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+    targets : Theano 2D tensor or 1D tensor
+        Either a vector of int giving the correct class index per data point
+        or a 2D tensor of one-hot encoding of the correct class in the same
+        layout as predictions (non-binary targets in [0, 1] do not work!)
```
Wasn't this meant for regression? You copied the description from a multi-class loss.

Sentient07 replied:
Ah! Yes, I will revert.

f0k commented on lasagne/objectives.py:
```diff
@@ -338,6 +338,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+    .. math:: L_\delta (diff) = \frac{diff^2}{2} & \text{if  } |diff| \le \
+                                \delta
+    .. math:: L_\delta (diff) & = & \delta (|diff| - \frac{\delta}{2} ),\
+                                     &\text{else}
```
Haven't checked what this renders like, but shouldn't this use a single `.. math::` statement? http://www.sphinx-doc.org/en/stable/ext/math.html#directive-math
It seems rendering cases with a one-sided bracket is not easily possible in Sphinx, so using two lines is a good workaround.

Sentient07 replied:
I used two here because there are two different expressions that this will return, based on the value of delta. To represent that, I used two math.

f0k replied:
Usually you'd use a single expression with a conditional bracket, but that's not easily possible in Sphinx (http://tex.stackexchange.com/questions/122407/writing-conditional-equations-with-braces-in-sphinx).
Using two lines seems like a good alternative, but you can write two lines in a single `.. math::` statement, as shown at http://www.sphinx-doc.org/en/stable/ext/math.html#directive-math. I guess that would be cleaner.

f0k commented on lasagne/objectives.py:
```diff
@@ -338,6 +338,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+        layout as predictions (non-binary targets in [0, 1] do not work!)
+    delta : scalar, default 1
+        This delta value is defaulted to 1, for SmoothL1Loss
+        described in Fast-RCNN paper[1].
```
Should be ` [1]_`, with an underscore in the end and a space before, if I recall the syntax correctly.

f0k commented on lasagne/objectives.py:
```diff
@@ -338,6 +338,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+    Notes
+    -----
+    This is an alternative to the Least Squared loss for
+    regression problems.
```
`least squares`, or better `squared error` because that's what the function is called in Lasagne.

f0k commented on lasagne/objectives.py:
```diff
@@ -338,6 +338,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+    Returns
+    -------
+    Theano 1D tensor
+        An expression for the item-wise huber loss.
```
Again, this was copied from a multi-class objective docstring. Copy it from the squared error instead (it should be a tensor of arbitrary dimensionality, and the element-wise loss).

f0k commented on lasagne/objectives.py:
```diff
@@ -338,6 +338,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+    diff = targets - predictions
+    ift = 0.5 * squared_error(targets, predictions)
+    iff = delta * (abs(diff) - delta / 2.)
+    return theano.tensor.switch(abs(diff) <= delta, ift, iff).sum()
```
Don't sum in the end, it should return the element-wise loss.

f0k commented on lasagne/objectives.py:
```diff
@@ -338,6 +338,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+           https://arxiv.org/pdf/1504.08083.pdf
+    """
+    predictions, targets = align_targets(predictions, targets)
+    diff = targets - predictions
```
Use `abs_diff = abs(targets - predictions)` instead. Maximizes reuse of Theano expressions, so Theano doesn't have to merge them later.

f0k commented on lasagne/tests/test_objectives.py:
```diff
@@ -153,6 +153,35 @@ def test_binary_hinge_loss(colvect):
 @pytest.mark.parametrize('colvect', (False, True))
+def test_huber_loss(colvect):
+    from lasagne.objectives import huber_loss
+    delta = [0.5, 1.0]
```
You can add `delta` to the test parameters and add another `parametrize`. Looks a little cleaner.

f0k commented on lasagne/objectives.py:
```diff
@@ -338,6 +338,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+    ----------
+    .. [1] Ross Girshick et al (2015):
+           Fast RCNN
+           https://arxiv.org/pdf/1504.08083.pdf
```
Can you also cite Huber, maybe as the first reference? https://en.wikipedia.org/wiki/Huber_loss#cite_note-1

f0k commented:
Oh, and I forgot, the new function should be included in `__all__` at the top of the file, and in the module docstring at the top of the file, and in `docs/modules/objectives.rst`.

Sentient07 commented:
I have added a new commit that addresses all the comments to the best of my knowledge. I haven't verified myself what sphinx renders. There seems to be some problem with my browser(safari). If the sphinx isn't fine, please let me know I will change it and try to build that on another system and push one more commit.
Thanks

f0k commented on lasagne/objectives.py:
```diff
@@ -86,6 +86,7 @@
     "aggregate",
     "binary_hinge_loss",
     "multiclass_hinge_loss",
+    "huber_loss",
```
Please also add it to the module docstring further above.

f0k replied:
And to the `.rst` file as I mentioned in https://github.com/Lasagne/Lasagne/pull/819#issuecomment-291599153. Thank you!

f0k commented on lasagne/objectives.py:
```diff
@@ -341,45 +342,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
+    .. math:: 
+        L_\delta (diff) = \frac{diff^2}{2} & \text{if  } |diff| \le \ delta \\
+
+        L_\delta (diff) & = & \delta (|diff| - \frac{\delta}{2} ),\ \\
```
The alignment characters (`&`) still seem off -- does this render correctly in Sphinx?

f0k commented on lasagne/objectives.py:
```diff
@@ -341,45 +342,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
-        or a 2D tensor of one-hot encoding of the correct class in the same
-        layout as predictions (non-binary targets in [0, 1] do not work!)
+        Ground truth to which the prediction is to be compared
+        with.
```
Please copy the `predictions` and `targets` from `binary_crossentropy`. This is still not correct.

Sentient07 replied:
I am sorry, it's a little unclear to me. You mean to say, i should be having predictions and targets in a single line separated by `,` similar to squared_error?

Sentient07 replied:
Something like this?
```
    Parameters
    ----------
    predictions,  targets : Theano tensors
```

f0k replied:
Yes, or you can also call it `a` and `b`, since the loss is completely symmetric. Your current docstring says they should be 1D or 2D tensors, which is too strict! They can be anything -- it's a drop-in replacement for the squared error.

f0k commented on lasagne/objectives.py:
```diff
@@ -341,45 +342,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
            Fast RCNN
            https://arxiv.org/pdf/1504.08083.pdf
+
+    .. [2] Huber, Peter et al (1964)
```
This line should have a colon `:` in the end. Sorry for the apparent nitpick, but note that the line break is only visible when viewing the docstring in Python, not in the Sphinx rendering. Also please add a period `.` after the title. Have a look at how this renders to be sure. (Last bullet point: https://github.com/Lasagne/Lasagne/blob/master/.github/PULL_REQUEST_TEMPLATE.md)

f0k replied:
Ah, didn't see the comment about your browser problem. Do you have another browser you can try? (Using another computer seems overkill!) Otherwise I can also check out your PR and test the documentation here, but of course it's easier if you can debug and fix things directly!

f0k commented on lasagne/objectives.py:
```diff
@@ -341,45 +342,50 @@ def multiclass_hinge_loss(predictions, targets, delta=1):
     """
     predictions, targets = align_targets(predictions, targets)
-    diff = targets - predictions
+    ab_diff = abs(targets - predictions)
```
I'd strongly prefer `abs_diff`.

Sentient07 replied:
It was a typo. I don't have anything against `abs_diff`, but yeah I understand the convention :)

f0k commented on lasagne/tests/test_objectives.py:
```diff
@@ -153,17 +153,17 @@ def test_binary_hinge_loss(colvect):
-        l1 = huber_loss(a, b, delta[0])
-        l2 = huber_loss(a, b, delta[1])
+        l1 = huber_loss(a, b, delta)
+        l2 = huber_loss(a, b, delta)
```
You're computing the same thing twice now. The idea of making `delta` a test parameter was to remove the duplication :)

Sentient07 replied:
Ah! Yeah, sorry. I rushed through that PR. I will check every comment again before the next commit.

Sentient07 commented:
I very much apologise for the delay. The two months were really hectic in seeking positions. I have addressed the comments that were left over. The sphinx now looks good in my local build. If there are anymore changes, I will make them right away! 
<img width="756" alt="screenshot at jun 01 01-24-30" src="https://cloud.githubusercontent.com/assets/6195312/26651447/53d708ec-466a-11e7-89b3-fe8b25bb919e.png">

f0k commented on lasagne/tests/test_objectives.py:
```diff
@@ -153,6 +153,30 @@ def test_binary_hinge_loss(colvect):
+    abs_diff = abs(x - y)
+    ift = 0.5 * abs_diff ** 2
+    iff = delta * (abs_diff - delta / 2.)
+    z = np.where(abdiff <= delta, ift, iff)
```
I'd merge this right away, only all your tests failed with `NameError: global name 'abdiff' is not defined`.

Sentient07 replied:
Sorry about that, I will fix it.

f0k commented:
Looks good now, can you please squash everything into the first commit?
